overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,466 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package common
+
+import (
+	"encoding/base32"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	ErrInvalidSnapLabel           = errors.New("the given label cannot be converted to snap")
+	ErrPermissionNotInList        = errors.New("permission not found in permissions list")
+	ErrPermissionsListEmpty       = errors.New("permissions list empty")
+	ErrUnrecognizedFilePermission = errors.New("file permissions mask contains unrecognized permission")
+)
+
+type Constraints struct {
+	PathPattern string   `json:"path-pattern,omitempty"`
+	Permissions []string `json:"permissions,omitempty"`
+}
+
+// ValidateForInterface returns nil if the constraints are valid for the given
+// interface, otherwise returns an error.
+func (c *Constraints) ValidateForInterface(iface string) error {
+	switch iface {
+	case "home", "camera":
+		// TODO: change to this once PR #13730 is merged:
+		// if err := ValidatePathPattern(c.PathPattern); err != nil {
+		//	return err
+		// }
+	default:
+		return fmt.Errorf("constraints incompatible with the given interface: %q", iface)
+	}
+	permissions, err := AbstractPermissionsFromList(iface, c.Permissions)
+	if err != nil {
+		return err
+	}
+	c.Permissions = permissions
+	return nil
+}
+
+// Match returns true if the constraints match the given path, otherwise false.
+//
+// If the constraints or path are invalid, returns an error.
+func (c *Constraints) Match(path string) (bool, error) {
+	// TODO: change to this once PR #13730 is merged:
+	// return PathPatternMatch(c.PathPattern, path)
+	return true, nil
+}
+
+// RemovePermission removes every instance of the given permission from the
+// permissions list associated with the constraints. If the permission does
+// not exist in the list, returns ErrPermissionNotInList.
+func (c *Constraints) RemovePermission(permission string) error {
+	origLen := len(c.Permissions)
+	for i := 0; i < len(c.Permissions); {
+		perm := c.Permissions[i]
+		if perm != permission {
+			i++
+			continue
+		}
+		copy(c.Permissions[i:], c.Permissions[i+1:])
+		c.Permissions = c.Permissions[:len(c.Permissions)-1]
+	}
+	if origLen == len(c.Permissions) {
+		return ErrPermissionNotInList
+	}
+	return nil
+}
+
+// ContainPermissions returns true if the constraints include every one of the
+// givne permissions.
+func (c *Constraints) ContainPermissions(permissions []string) bool {
+	for _, perm := range permissions {
+		if !strutil.ListContains(c.Permissions, perm) {
+			return false
+		}
+	}
+	return true
+}
+
+type OutcomeType string
+
+const (
+	OutcomeUnset OutcomeType = ""
```

Comment:
Nitpick: maybe a value similar to this would make error messages a bit clearer? Otherwise an error message could be `must be "allow" or "deny": ""` 
```suggestion
	OutcomeUnset OutcomeType = "(unset)"
```

Reply:
I like your reasoning, but I would like to leave the Unset values as `""`, since when patching existing rules, only modified fields need to be included (others are omitempty), and by using `""` the default value when unmarshalling from an omitted outcome or lifespan type are the Unset variants.

Reply:
That's a good point. Thanks